packages/go_router/lib/src/misc/stateful_navigation_shell.dart

```diff
@@ -0,0 +1,185 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+import 'package:flutter/widgets.dart';
+
+import '../configuration.dart';
+import '../typedefs.dart';
+
+/// [InheritedWidget] for providing a reference to the closest
+/// [StatefulNavigationShellState].
+class InheritedStatefulNavigationShell extends InheritedWidget {
+  /// Constructs an [InheritedStatefulNavigationShell].
+  const InheritedStatefulNavigationShell({
+    required super.child,
+    required this.state,
```

Comment:
why not take in the StatefulShellRouteState instead?

Reply:
I was back and forth with this, and I actually ended up using `StatefulNavigationShellState` partly due to the fact that `StatefulShellRouteState` was generated on the fly via a getter. But after this (and above) feedback, my thinking now is to instead keep a `StatefulShellRouteState` instance within `StatefulNavigationShellState`, i.e. instead of `_currentIndex` and `_childRouteState`. That way there might not be the same need to implement the equality operator in `StatefulShellRouteState`, i.e. identity equality would suffice. But of course, we could still implement basic/shallow equality in `StatefulShellRouteState` (might save some unnecessary rebuilds): 

```dart
  @override
  bool operator ==(Object other) {
    if (other == this) {
      return true;
    }
    if (other is !StatefulShellRouteState) {
      return false;
    }
    return other.route == route && other.index == index;
  }

  @override
  int get hashCode => Object.hash(route, index);
```